dev-libs/aws-sdk-cpp: version bump to 1.7.178

Yamakuzure opened the pull request:
This is a folow up to https://github.com/gentoo/gentoo/pull/10072

Besides the version bump, three dependencies had to be added:
 * dev-libs/aws-c-common: Core c99 package for AWS SDK for C
 * dev-libs/aws-c-event-stream: amazon evenstream C99 implementation
 * dev-libs/aws-checksums: HW accelerated CRC32c and CRC32 with fallbacks

I have not yet tested everything as I have very limited time right now. Luckily I need this at work, so I can fix things quickly.
Unfortunately I needed the new SDK fast, so please have an extra hard look at the changes, I am certain that I haven't done everything 100% correctly. Regardless what repoman tells me.

Thanks in advance!

Closes: https://bugs.gentoo.org/666494
Closes: https://bugs.gentoo.org/665540
Signed-off-by: Sven Eden <yamakuzure@gmx.net>

Yamakuzure commented:
@ambasta : As this is based on your work on bumping to 1.6.23 at #10072 , could you please have a look whether this looks ok for you?

@jonasstein : You filed the original bump request at https://bugs.gentoo.org/666494 , could you please have a look?

I have no merge authority and need a proxy, too.

juippis commented on dev-libs/aws-checksums/aws-checksums-0.1.3.ebuild:
```diff
@@ -0,0 +1,26 @@
+
+IUSE="test"
+
+DEPEND="dev-util/ninja"
```
inheriting cmake-utils.eclass will grant you the ninja dependency here. So you can just remove DEPEND and RDEPEND from this ebuild, and from the other similar ebuilds in this PR.

Yamakuzure replied:
Got it!
And while on it, I also found an error with DEPEND/RDEPEND in the aws-sdk-cpp ebuild.

juippis commented on dev-libs/aws-sdk-cpp/aws-sdk-cpp-1.7.160.ebuild:
```diff
@@ -0,0 +1,95 @@
+KEYWORDS="~amd64 ~x86"
+
+MODULES=(
+	access-management acm acm-pca alexaforbusiness amplify apigateway
```
Seriously... I know this wasn't done by you, but you could make it better. Is it possible to pick few *important* flags, keep them and enable/disable **all** the others with a single USE flag (eg `extra-modules`?)
This would also make it super easy to check for required disk space and memory (bug #665540), since it'd be dependant on a single USE flag.

Yamakuzure replied:
Maybe some can be put together, like putting all kinesis modules together.

But generally this is a very difficult task. There are no important flags but `core` which is enabled by default.

Example: Someone who produces Lambdas will most probably not need anything else, unless their lambdas want to talk to, let's say, S3 or something.

However, you are, of course, right. The list of modules has become extremely vast! And most of these names only mean something for people who actually have to work with them.

Some of the modules depend on each other, and the build system will enable them. But some sort of catalog to reduce the (unholy!) amount of flags right now would be great.

I'll see what I can map during the week-end.

Maybe there are some AWS specialists here who can pipe in and give their 2 cents?

Yamakuzure commented:
Added [nomerge] to the title.
I really want to try to reduce this huge number of modules/flags before this PR gets merged.
Any help and suggestions are highly appreciated!

Yamakuzure commented:
Damn...
Note to self: When working on **two** systems, make sure to pull from origin and update your dev branch, *BEFORE* resetting hard on gentoo/master !

Yamakuzure commented:
This is vast!

See: https://docs.aws.amazon.com/index.html

There are many services, however, that we can put together. But it'll take a while.

Yamakuzure commented:
Just a quick update:
It is not possible to scratch the list together by a large sum. But a few modules are really just parts of others.
I fear the list will end up being still very long. But while I am analyzing what can be done, I also document the USE flags so that you don't have to be an AWS Guru to know what's what.

However, this is much work and I guess I will need at least the remainder of this week. Maybe I won't be able to present a result before next week. So I have to ask for a little bit of patience.

juippis commented:
It's fine from our part, thanks for the status update and nice to see you're doing progress. 

Is there some way to sort these modules by popularity in some Amazon site? I'd still have max 10 modules as USE flags, and as much as possible under "extras" or something like that.

Yamakuzure commented:
I don't know whether such statistics like module popularity even exist. 

Well, I am halfway through and have killed 20 flags and added 2 new (*). I hope the second half provides some more opportunities to attach modules to others.

(*) The new are:
- cloud-dev : Which activates all development in cloud modules like cloud9 and xray
- cognito : As it really consists of three modules that make no sense to split.

Yamakuzure commented:
@ambasta @jonasstein @juippis 
It is done!
This is the first real take, but it is not yet meant for merging. First the following two points remain:

1. I *think* there are more modules that can be put together, but what I could fathom by waltzing through the documentation and the SDK headers themselves, this is it for now. Maybe you can detect something worth merging?
2. I am not too happy with the way the modules are put together. But I failed to come up with something more elegant than the if-elif-fi list.

The module list will be cleaned up and aligned when I'll process you feedback.

Thank you very much for your help!

juippis commented:
`cloud` for everything cloud* related? 
I'm more after this:
https://devmanual.gentoo.org/general-concepts/use-flags/index.html
Check the "when not to use USE flags" part. 

>A similar case can be made for packages having multiple conditional programs or modules. Whenever this results in a large number of USE flags that would force the user to spend a lot of time choosing compatible flags and possibly rebuilding after incomplete choices, consider reducing the use of flags to those programs or modules that have external dependencies and/or long build times. The rest of them should be built unconditionally instead, or controlled by a flag such as minimal. 

So I'd keep what you have, perhaps add the `cloud` USE if it makes sense, then put *everything* else under one USE flag. Or do the opposite way, if `minimal` is not set, enable everything beside those controlled by USE flags.

You should use array form to control your build targets. The ebuild will be long and ugly, but it won't hinder user experience.

Yamakuzure commented:
Forgot to mention yesterday:

Added 7 flags
Removed 93 flags
Reduced by: 86 flags 

I think, although the list is still long, a good outcome.

Yamakuzure commented:
> `cloud` for everything cloud* related?

That would be everything, as AWS is about stuff in the cloud. ;-)

> I'm more after this:
> https://devmanual.gentoo.org/general-concepts/use-flags/index.html
> Check the "when not to use USE flags" part.

I am fully aware of that part, but the only way to circumvent a long list of USE flags is to turn the modules into an expandable member of USE_EXPAND.
 (Like it seemed to be prepared for with the aws_targets_ prefix.)

The crux of the "When not to use USE flags" part is, that it says to try to reduce the flags when the large amount forces the user to spend a lot of time choosing compatible flags.

1. All flags are "compatible" in a way that the build system will not allow to chose otherwise. No modules block others.
2. It is AWS we are talking about. The most stuff isn't even related to each other. Just look at "groundstation", "macie" and "route53". How could they be combined?

The user ***will*** have to spend **much** time studying dozens, if not hundreds, of pages of documentation before they even know where to start. Maybe I just have too much respect for this vast field of cloud computing, but the first time I took a look at the list of modules in the SDK, I felt like I've been hit with a hammer.

Look at the list now, and the explanation in the metadata.xml. Do you really want to get rid of that documentation? Please compare with `equery u dev-libs/aws-sdk-cpp-1.5.17`.

Keeping unrelated modules apart and then at least document the flags isn't an obstacle, it is an assistance for the user.

I have just copied everything into my `/usr/local/portage` and ran `equery u dev-libs/aws-sdk-cpp-1.7.171` and the output is really nice. Everything is explained. The formatting from the `metadata.xml` is gone, sure, but it'll be a big help anyway.

Before you think otherwise, I fully concur with you that the list of USE flags must be reduced further. A nice side effect of the `equery` output is, however, that the listing is much better comprehensible than the `metadata.xml`.
As an example, I think `appmesh` can be activated by `monitoring`. And I bet there will be man other modules that can be combined. Maybe even with new meta-flags.

Still there will be many USE flags. 116 are there now, including the regular ones. Even after combining more flags into *logical* groups of *related* modules there will be man left. Maybe 80 or even 90.
So an expandable might seem to be the best choice, but `/usr/portage/profiles/base/make.defaults` clearly states:
````
# Env vars to expand into USE vars.  Modifying this requires prior
# discussion on gentoo-dev@lists.gentoo.org.
````
So this won't happen soon. And I daresay we first have to have a final list of flags before bringing those forward to the dev list, haven't we?

(Sorry, I completely wandered off...)

> So I'd keep what you have, perhaps add the `cloud` USE if it makes sense, then put _everything_
> else under one USE flag. Or do the opposite way, if `minimal` is not set, enable everything beside
> those controlled by USE flags.

I think I didn't really understand what you mean here. As I said, there is no really easy way to define "everything else". 
I mean, can you, of the top of your head, explain what the modules "cur", "dms", "eks", "kms", "sdb", "sms" and "ssm", do?
Would you think someone just starting would even have an idea that "cur" stands for "Cost and Usage Report Service", and that "sms" hasn't got anything to do with messages but means "Server Migration Service"?

Okay, I admit it, it is of course also some sort of egoism on my part, as this flag documentation helps me a great deal. 

> You can wait for bug reports to make some modules optional. Also, if you know there are some modules that take a lot of time to build, make them optional.

Finding out which modules need long build times all by themselves won't be easy. Although I could write a script that emerges the sdk over and over again with alternating USE flags, timing the merge.

I think I will do that, thanks for the inspiration!
While looking at the `equery u` output I have an idea that many of the *elastic* USE flags, some of which do not even have the name 'elastic' in them, might be combined to one really large meta-flag.
And I already have some meta-flags.
Knowing what causes long build times may be valuable.

> You should use array form to control your build targets. The ebuild will be long and ugly, but it won't
> hinder user experience.

The list of modules is already an array. Or do you mean the local mybuildtargets string?

Thank you very much for your input! And please pardon me if I come over somewhat stubborn.

juippis commented:
> > `cloud` for everything cloud* related?
> 
> That would be everything, as AWS is about stuff in the cloud. ;-)

I just meant every module that has the word `cloud` in it :P 

> 
> I think I didn't really understand what you mean here. As I said, there is no really easy way to define "everything else".

I mean, it's very easy to define but up to you to support the idea. Simply put, have a USE flag which controls the state of say 100 modules. "build-everything", or "minimal", or something. 

> 
> Okay, I admit it, it is of course also some sort of egoism on my part, as this flag documentation helps me a great deal.
> 

I agree the documentation is good. 


> Knowing what causes long build times may be valuable.
> 

I think so too, but don't overdo it. It's optional...

> 
> The list of modules is already an array. Or do you mean the local mybuildtargets string?

Yes, you can just append them to the array. 

> 
> Thank you very much for your input! And please pardon me if I come over somewhat stubborn.

Enthusiasm is important ;)

@ambasta any word about where this is heading?

Yamakuzure commented:
> > > `cloud` for everything cloud* related?
> > 
> > That would be everything, as AWS is about stuff in the cloud. ;-)
> 
> I just meant every module that has the word `cloud` in it :P

Ah! Alright!
Well, that was my first idea, too, until I saw that the same keyword can mean something completely different.
Example:
- Elastic Compute Cloud (EC2) -> Server
- Elastic Container Service (ECS) -> Docker
- Elastic Kubernetes Service (EKS) -> Kubernetes
However, I currently *am* trying to tie some more modules together.

> > I think I didn't really understand what you mean here. As I said, there is no really easy way to define "everything else".
> 
> I mean, it's very easy to define but up to you to support the idea. Simply put, have a USE flag which controls the state of say 100 modules. "build-everything", or "minimal", or something.

This would be simple if there they were related. I do what I can, but it is difficult to explain why a module for satellite ground control should be activated by the same flag like another module for a managed graph database service.

Maybe once I have finished "condensing" some patterns evolve to 'categorize' certain modules even further. 

> I agree the documentation is good.

Thanks. :smile: 

> > Knowing what causes long build times may be valuable.
> 
> I think so too, but don't overdo it. It's optional...

Well, the biggest one was EC2, followed closely by cloud-iot, which is a meta-flag. Unfortunately my analysis wasn't really telling in the end, as the metaflags came on top, which is logical, and then the modules almost linearly produced smaller and smaller build times. Apart from EC2 there were no spikes.

> > The list of modules is already an array. Or do you mean the local mybuildtargets string?
> 
> Yes, you can just append them to the array.

In the end I will need a string, so what would be the advantage of
´´´´
mymodules+=( foo )
````
over
````
mymodules+=";foo"
````
when the latter is already the form needed?

However, an associative array *would* be helpful, but I guess that would be not allowed in an ebuild, right? If it were I could make the module names the keys and wouldn't need to check against doublets.

From the logical view I have finished, but not yet pushed. There is another clue I'd like to follow, however:

Some modules enable others the depend on or complement. I will check whether there are "siblings", thus modules where module A enables module B and vice versa. Those can be combined even if they are not seemingly related.

Maybe this reduces the list enough to really end up with more special and/or obscure modules that then can be combined although not related.

Yamakuzure commented:
> Some modules enable others the depend on or complement. I will check whether there are "siblings", thus modules where module A enables module B and vice versa. Those can be combined even if they are not seemingly related.
> 
> Maybe this reduces the list enough to really end up with more special and/or obscure modules that then can be combined although not related.

Okay, this is a dead end.

However, th in-tree ebuild lists 151 modules, while the new on (once pushed) lists 101.
So there is at least some improvement. The number of modules got reduced by a third.

Yamakuzure commented:
@juippis :
> I mean, it's very easy to define but up to you to support the idea. Simply put, have a USE flag which controls the state of say 100 modules. "build-everything", or "minimal", or something.

After having condensed and combined what I could, I made a list of 'other' modules that *might* be rarely needed. They are now combined under the "other" USE flag.

I have added a note to the USE flag description that users wanting to single any of those 'other' modules out shall open a bug for it.

@ambasta @jonasstein  : What do you guys think?

I have also moved to 1.7.178, as this is the last revision with enhancements before the SDK introduces a new service (Amazon QLDB)

It is my belief that we should stop here and first get the ebuilds into the tree before proceeding further. The AWS SDK moves incredibly fast anyway.

juippis commented:
I'll start the test runs. One question: Would you like to become a proxy-maintainer for aws-sdk-cpp seeing you've done a lot of work for it, and in future if you are the maintainer we can merge it faster.

juippis commented on dev-libs/aws-c-common/aws-c-common-0.4.3.ebuild:
```diff
@@ -0,0 +1,26 @@
+
+IUSE="test"
+
+DEPEND=""
```
You can remove these empty variables from the ebuild if you wish so.

Yamakuzure replied:
Check!

juippis commented on dev-libs/aws-c-common/metadata.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+	<longdescription>
```
Please follow the metadata skel and put longdescription after maintainers. I'm asking to make this change because there are a lot of scripts that crawl the metadata info, and while I doubt this will break anything, let's make sure we are consistent with other in-tree packages.

Yamakuzure replied:
Check

juippis commented on dev-libs/aws-c-event-stream/aws-c-event-stream-0.1.1.ebuild:
```diff
@@ -0,0 +1,26 @@
+
+IUSE="test"
+
+DEPEND=""
```
Same here.

Yamakuzure replied:
Check

juippis commented on dev-libs/aws-checksums/aws-checksums-0.1.3.ebuild:
```diff
@@ -0,0 +1,26 @@
+
+IUSE="test"
+
+DEPEND=""
```
And here.

Yamakuzure replied:
Check

juippis commented on dev-libs/aws-checksums/metadata.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+	<longdescription>
```
And here (metadata).

Yamakuzure replied:
Check

juippis commented on dev-libs/aws-sdk-cpp/aws-sdk-cpp-1.7.178.ebuild:
```diff
@@ -0,0 +1,147 @@
+REQUIRED_USE="${PYTHON_REQUIRED_USE}"
+
+DEPEND="
+	http?	( net-misc/curl:= )
```
I know it looks better, but let's follow the default style and keep one space here between `? (` :)

Yamakuzure replied:
*meh* I used one tab. Indentation is tabbed any way.
but ... *sigh ... Check!

juippis commented on dev-libs/aws-sdk-cpp/aws-sdk-cpp-1.7.178.ebuild:
```diff
@@ -0,0 +1,147 @@
+
+DEPEND="
+	http?	( net-misc/curl:= )
+	polly?	( media-sound/pulseaudio )
```
Here too.

juippis commented on dev-libs/aws-sdk-cpp/aws-sdk-cpp-1.7.178.ebuild:
```diff
@@ -0,0 +1,147 @@
+DEPEND="
+	http?	( net-misc/curl:= )
+	polly?	( media-sound/pulseaudio )
+	ssl?	( dev-libs/openssl:0 )
```
And here. Any idea about libressl compatibility? openssl has a soname, so if this gets linked during compilation (its in DEPEND after all) you need to add a subslot binder, `:=`.

Yamakuzure replied:
The build system, as far as I can tell, only detects that it is there and uses it as a default provider for SSL.

This might imply that the SDK *could* be compatible to libressl, but I honestly do not know.

However, I have enabled support for it and if a user with a libressl system finds out that the assumption is wrong, it can be patched to make it work or the support can be removed.

juippis commented on dev-libs/aws-sdk-cpp/aws-sdk-cpp-1.7.178.ebuild:
```diff
@@ -0,0 +1,147 @@
+	sys-libs/zlib
+"
+RDEPEND="
+		${DEPEND}
```
One tab too many.

Yamakuzure replied:
Ooops! :scream: 

Check!

juippis commented on dev-libs/aws-sdk-cpp/aws-sdk-cpp-1.7.178.ebuild:
```diff
@@ -0,0 +1,147 @@
+"
+RDEPEND="
+		${DEPEND}
+		${PYTHON_DEPS}
```
And here.

juippis commented:
aws-c-event-stream won't compile with test enabled, 
```
82% tests passed, 2 tests failed out of 11

Total Test time (real) =   0.04 sec

The following tests FAILED:
	  8 - test_streaming_decoder_incoming_no_op_valid_single_message (Child aborted)
	 11 - test_streaming_decoder_incoming_multiple_messages (Child aborted)
Errors while running CTest
```
```
8/11 Testing: test_streaming_decoder_incoming_no_op_valid_single_message
8/11 Test: test_streaming_decoder_incoming_no_op_valid_single_message
Command: "/tmp/portage/dev-libs/aws-c-event-stream-0.1.1/work/aws-c-event-stream-0.1.1_build/tests/aws-c-event-stream-tests" "test_streaming_decoder_incoming_no_op_valid_single_message"
Directory: /tmp/portage/dev-libs/aws-c-event-stream-0.1.1/work/aws-c-event-stream-0.1.1_build/tests
"test_streaming_decoder_incoming_no_op_valid_single_message" start time: Sep 17 17:49 -00
Output:
----------------------------------------------------------
Fatal error condition occurred in /tmp/portage/dev-libs/aws-c-common-0.4.3/work/aws-c-common-0.4.3/source/common.c:72: size != 0
Exiting Application
0x00007ff91df36b69: ?? ??:0
0x00007ff91df39ba9: ?? ??:0
0x0000000000005b0c: s_decoder_test_on_prelude_received at message_streaming_decoder_test.c:?
0x0000000000002729: s_verify_prelude_state at event_stream.c:?
0x00007ff91df54f0e: ?? ??:0
0x0000000000007499: s_test_streaming_decoder_incoming_no_op_valid_single_message_fn at message_streaming_decoder_test.c:?
0x00000000000078ce: test_streaming_decoder_incoming_no_op_valid_single_message at :?
0x0000000000002450: main at :?
0x00007ff91dd57f1b: ?? ??:0
0x00005612cce9d65a: ?? ??:0
<end of output>
Test time =   0.03 sec
----------------------------------------------------------
Test Failed.
"test_streaming_decoder_incoming_no_op_valid_single_message" end time: Sep 17 17:49 -00
"test_streaming_decoder_incoming_no_op_valid_single_message" time elapsed: 00:00:00

11/11 Testing: test_streaming_decoder_incoming_multiple_messages
11/11 Test: test_streaming_decoder_incoming_multiple_messages
Command: "/tmp/portage/dev-libs/aws-c-event-stream-0.1.1/work/aws-c-event-stream-0.1.1_build/tests/aws-c-event-stream-tests" "test_streaming_decoder_incoming_multiple_messages"
Directory: /tmp/portage/dev-libs/aws-c-event-stream-0.1.1/work/aws-c-event-stream-0.1.1_build/tests
"test_streaming_decoder_incoming_multiple_messages" start time: Sep 17 17:49 -00
Output:
----------------------------------------------------------
Fatal error condition occurred in /tmp/portage/dev-libs/aws-c-common-0.4.3/work/aws-c-common-0.4.3/source/common.c:72: size != 0
Exiting Application
0x00007f976907ab69: ?? ??:0
0x00007f976907dba9: ?? ??:0
0x0000000000005b0c: s_decoder_test_on_prelude_received at message_streaming_decoder_test.c:?
0x0000000000002729: s_verify_prelude_state at event_stream.c:?
0x00007f9769098f0e: ?? ??:0
0x0000000000005d93: s_test_streaming_decoder_incoming_multiple_messages_fn at message_streaming_decoder_test.c:?
0x0000000000007f5e: test_streaming_decoder_incoming_multiple_messages at :?
0x0000000000002450: main at :?
0x00007f9768e9bf1b: ?? ??:0
0x0000557af740965a: ?? ??:0
<end of output>
Test time =   0.03 sec
----------------------------------------------------------
Test Failed.
"test_streaming_decoder_incoming_multiple_messages" end time: Sep 17 17:49 -00
"test_streaming_decoder_incoming_multiple_messages" time elapsed: 00:00:00
----------------------------------------------------------

End testing: Sep 17 17:49 -00
```

juippis commented:
^ but I'm willing to merge this after the reviewed styling issues is taken care of, you can restrict the non-working tests for now and fix them for a next version bump / in your free time with new PR.

Honestly it's going to be a maintenance hell, but good work so far :)

Yamakuzure commented:
> aws-c-event-stream won't compile with test enabled,
> 
> ```
> 82% tests passed, 2 tests failed out of 11
> 

Fixed in aws-c-event-stream-0.1.3 :smile: 

So I'll bump it

Yamakuzure commented:
@juippis : Thanks you very much for your help!

juippis commented:
You too, thanks for patience and good work! Feel free to add yourself as a proxy-maintainer for this package. So next time you make pull requests for these packages, it labels the PR as "self-maintained" and it shows up higher in our queue.

Yamakuzure commented:
> Feel free to add yourself as a proxy-maintainer for this package. So next time you make pull requests for these packages, it labels the PR as "self-maintained" and it shows up higher in our queue.

I will do that! :+1:
